rfcs/20190411-tensorboard-improved-plugin-ext.md

```diff
@@ -0,0 +1,151 @@
+
+#### Option 3: iframe
+**Pro**
+- ease of development. Can bring your own version of Polymer, React, and etc...
+- encapsulation and containment of plugins accidentally doing wrong things
+
+**Con**
+- does not work in Colab
+- generally larger binaries
+- Cannot reuse components across plugins
+- can lead to an awkward UX: e.g., different version of a visualization library have different UI/UX
+
+When a plugin is activated by a user, TensorBoard will dynamically create an iframe and point the iframe to the endpoint that plugin defined as an entry point.
+
+The solution is the most favorable because it gives plugin developers a lot of freedom in terms of frameworks. With global state disjoint from others, each plugin binary can incorporate as many dependencies as it would like to use without worrying about the correctness. This is quite appealing especially for Polymer-based applications because Polymer uses a global state, `CustomElementRegistry` (in v1, `document.registerElement`), and is not possible to load components with the same component name more than once. Loading plugins in its own separate frame will let plugin authors develop with ease and potentially have a standalone widgets for use outside of TensorBoard.
+
+It must be clearly stated that the use of global state is not unique to Polymer. Although JavaScript bundles use Immediately Invoked Function Expression (IIFE) and use closure to encapsulate dependencies, a bundler like Browserify or Webpack can uses global states when shimming CommonJS `window.require` and use non-unique identifier for a global module.
+
+Despite the benefits, iframe can be difficult to work with under Colab. Colab, running a VM under google.com domain, makes sure ports opened by user instantiated processes are not accessible publicly. Under this limitation, TensorBoard operates using a ServiceWorker based network proxy installed by Colab that routes a request to _localhost_ to an endpoint running in VM. This proxy is mostly transparent from TensorBoard as long as we make requests to the localhost.
+
+A difficulty arises with how ServiceWorker behave with respect to an iframe. A ServiceWorker can only intercept requests that originates from a document within its scope. A ServiceWorker, derived from its module source, is associated to a scope (for instance, if the ServiceWorker module originates from tensorflow.org/foo/bar/sw.js, it will have a scope of "tensorflow.org/foo/bar" and it can only intercept request from a document whose host starts with "tensorflow.org/foo/bar/"). Now, to render an iframe whose content source is backed by an endpoint running in a VM, we need to go through the Colab proxy by setting the source of the iframe as "https://localhost/*" (untrue; will be discussed below). This, though, is a problem to a request that originates within the iframe -- since iframe's origin is localhost, only ServiceWorkers with scope of "localhost" will be active and Colab's proxy will not be effective.
```

Comment:
> "https://localhost/*"

nit: use backtics to surround this. Or all the font becomes italic after this point (thanks to markdown renderer....)

Reply:
I am failing to follow :(

![image](https://user-images.githubusercontent.com/2547313/56054286-5266c100-5d0b-11e9-9f25-32182a675c43.png)